### PR TITLE
Use IUser->getEmailAddress and join the tables activity_mq and accoun…

### PR DESCRIPTION
**Pull request by DeepDiver1975**
…ts to get the user's email address right away

requires https://github.com/owncloud/core/pull/27205

### CODE REVIEW

**Comment by butonic** [comment]
👍 note that this tightly couples activity with the accounts table from core.

**Comment by DeepDiver1975** [comment]
> :+1: note that this tightly couples activity with the accounts table from core.

I know - but because we ship activity tighly coupled to code this is acceptable from my pov

**Comment by PVince81** [comment]
Switched back to master branch to rerun tests.

Would be good to redo a quick manual test to confirm it still works correctly.

**Comment by PVince81** [comment]
Tested, works :+1:

**Comment by PVince81** [comment]
This is critical, we MUST have this merged otherwise activity emails will stop working in 10.0.

Let's hope CI will pass now, I just rebased.

**Comment by PVince81** [comment]
```
1) OCA\Activity\Tests\MailQueueHandlerTest::testGetAffectedUsers with data set #0 (null, array('user2', 'user1', 'user3'), array())

Failed asserting that two arrays are equal.

--- Expected

+++ Actual

@@ @@

 Array (

-    0 => 'user2'

-    1 => 'user1'

-    2 => 'user3'

 )

/home/travis/build/owncloud/core/apps/activity/tests/MailQueueHandlerTest.php:170

/home/travis/build/owncloud/core/apps/activity/vendor/phpunit/phpunit/src/TextUI/Command.php:188

/home/travis/build/owncloud/core/apps/activity/vendor/phpunit/phpunit/src/TextUI/Command.php:118

2) OCA\Activity\Tests\MailQueueHandlerTest::testGetAffectedUsers with data set #1 (5, array('user2', 'user1', 'user3'), array())

Failed asserting that two arrays are equal.

--- Expected

+++ Actual

@@ @@

 Array (

-    0 => 'user2'

-    1 => 'user1'

-    2 => 'user3'

 )

/home/travis/build/owncloud/core/apps/activity/tests/MailQueueHandlerTest.php:170

/home/travis/build/owncloud/core/apps/activity/vendor/phpunit/phpunit/src/TextUI/Command.php:188

/home/travis/build/owncloud/core/apps/activity/vendor/phpunit/phpunit/src/TextUI/Command.php:118

3) OCA\Activity\Tests\MailQueueHandlerTest::testGetAffectedUsers with data set #2 (3, array('user2', 'user1', 'user3'), array())

Failed asserting that two arrays are equal.

--- Expected

+++ Actual

@@ @@

 Array (

-    0 => 'user2'

-    1 => 'user1'

-    2 => 'user3'

 )

/home/travis/build/owncloud/core/apps/activity/tests/MailQueueHandlerTest.php:170

/home/travis/build/owncloud/core/apps/activity/vendor/phpunit/phpunit/src/TextUI/Command.php:188

/home/travis/build/owncloud/core/apps/activity/vendor/phpunit/phpunit/src/TextUI/Command.php:118

4) OCA\Activity\Tests\MailQueueHandlerTest::testGetAffectedUsers with data set #3 (2, array('user2', 'user1'), array('user3'))

Failed asserting that two arrays are equal.

--- Expected

+++ Actual

@@ @@

 Array (

-    0 => 'user2'

-    1 => 'user1'

 )

/home/travis/build/owncloud/core/apps/activity/tests/MailQueueHandlerTest.php:170

/home/travis/build/owncloud/core/apps/activity/vendor/phpunit/phpunit/src/TextUI/Command.php:188

/home/travis/build/owncloud/core/apps/activity/vendor/phpunit/phpunit/src/TextUI/Command.php:118

5) OCA\Activity\Tests\MailQueueHandlerTest::testGetAffectedUsers with data set #4 (1, array('user2'), array('user1', 'user3'))

Failed asserting that two arrays are equal.

--- Expected

+++ Actual

@@ @@

 Array (

-    0 => 'user2'

 )

/home/travis/build/owncloud/core/apps/activity/tests/MailQueueHandlerTest.php:170

/home/travis/build/owncloud/core/apps/activity/vendor/phpunit/phpunit/src/TextUI/Command.php:188

/home/travis/build/owncloud/core/apps/activity/vendor/phpunit/phpunit/src/TextUI/Command.php:118

6) OCA\Activity\Tests\MailQueueHandlerTest::testSendEmailToUser

Failed asserting that two arrays are equal.

--- Expected

+++ Actual

@@ @@

 Array (

-    0 => Array (...)

 )

/home/travis/build/owncloud/core/apps/activity/tests/MailQueueHandlerTest.php:247

/home/travis/build/owncloud/core/apps/activity/vendor/phpunit/phpunit/src/TextUI/Command.php:188

/home/travis/build/owncloud/core/apps/activity/vendor/phpunit/phpunit/src/TextUI/Command.php:118
```

**Comment by PVince81** [comment]
I don't understand the failures. The code seems to only select the "email" value additionally so the returned affected users should stay the same. And yet it seems to return one more user:
```
1) OCA\Activity\Tests\MailQueueHandlerTest::testGetAffectedUsers with data set #0 (null, array('user2', 'user1', 'user3'), array())
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
     0 => 'user2'
     1 => 'user1'
-    2 => 'user3'
 )

/srv/www/htdocs/owncloud/apps3/activity/tests/MailQueueHandlerTest.php:170
```

@DeepDiver1975 any clues ?

**Comment by PVince81** [comment]
Found the problem: the new SQL query joins with the oc_accounts table and expects an entry to exist there. Since the test users weren't created in the accounts table, the query returns no results.

The test could be fixed by having proper entries in the accounts table.

With this new query there is a risk that the mail queue never gets cleared in case a user was deleted or a user cleared their email address while there were still entries to process. We might need special handling to delete such mail queue entries automatically. Normally entries are only deleted after processing.

Here I'd suggest to do a LEFT JOIN so we still receive the mail queue entries, but later we can decide to not send an email if the address is empty or user does not exist and still delete said entries.

I'll try and make the suggested changes:

- [ ] properly setup users in MailQueueHandlerTest to have entries in oc_accounts for old tests to pass
- [ ] proper handling for deleted users or users who removed their email address while entries exist in the queue

**Comment by PVince81** [comment]
@DeepDiver1975 please review my last commits

**Comment by PVince81** [comment]
- [ ] TODO: also add to $processedUsers if no email address is set

**Comment by PVince81** [comment]
@DeepDiver1975 I'm done post-pushing, please revieww

File: lib/AppInfo/Application.php [hunk 1]
```diff
@@ -177,6 +177,13 @@ public function __construct (array $urlParams = array()) {
 			$user = $server->getUserSession()->getUser();
 			return ($user) ? $user->getUID() : '';
 		});
+		$container->registerService('CurrentUser', function(IContainer $c) {
```
Comment (jvillafanez) [inline lib/AppInfo/Application.php]: Use core's ""getUserSession" service instead of creating a new one?

**Comment by jvillafanez** [comment]
Code-wise this is fine 👍

**Pull request merged**